src/3d/qgstessellatedpolygongeometry.cpp

```diff
@@ -78,6 +80,58 @@ QgsTessellatedPolygonGeometry::QgsTessellatedPolygonGeometry( bool _withNormals,
   }
 }
 
+static bool intersectionTriangles( const QByteArray &vertexBuf, const int &stride, const QgsRayCastingUtils::Ray3D &r, const QMatrix4x4 &worldTransform, QVector3D &intPt, int &triangleIndex )
+{
+  const float *vertices = reinterpret_cast<const float *>( vertexBuf.constData() );
+
+  const int vertexCount = vertexBuf.size() / stride;
+  const int triangleCount = vertexCount / 3;
+
+  QVector3D intersectionPt, minIntersectionPt;
+  float distance;
+  float minDistance = -1;
+
+  const int vertexSize = stride / sizeof( float );
+  const int triangleSize = 3 * vertexSize;
+  for ( int i = 0; i < triangleCount; ++i )
+  {
+    const int v0 = i * triangleSize, v1 = i * triangleSize + vertexSize, v2 = i * triangleSize + vertexSize + vertexSize;
+
+    const QVector3D a( vertices[v0], vertices[v0 + 1], vertices[v0 + 2] );
+    const QVector3D b( vertices[v1], vertices[v1 + 1], vertices[v1 + 2] );
+    const QVector3D c( vertices[v2], vertices[v2 + 1], vertices[v2 + 2] );
+
+    const QVector3D tA = worldTransform * a;
+    const QVector3D tB = worldTransform * b;
+    const QVector3D tC = worldTransform * c;
```

Comment:
it may be worth adding a note that currently the worldTransform only has vertical offset, so these transforms could be avoided by transforming the ray instead...

Reply:
fixed